### PR TITLE
CommunityBridge Project by Yash Varshney

### PR DESCRIPTION
**Pull request by Yash-Varshney**
Hello Everyone, I am Yash Varshney and this P.R. is regarding the community bridge mentorship programme. This pr contains all the four task done by me, previously I have made the p.r. for two tasks but this one contains all the task together. This project updates the python tool with v2.2 (although not completely) and introduces a CLI tool for parser and convertor. I have and will be raising some issues regarding future work needed to do. The next task will be to update the python tool with v3.x of python.
It was a great learning experience for me for the last 3 months and I will further be working with spdx-community.
This P.R. is open for review.

- [x]  Passing all the tests.

- [ ]  Review needed.

Signed-off-by: Yash Varshney <b18038@students.iitmandi.ac.in>

### CODE REVIEW

**Comment by Yash-Varshney** [comment]
ping : @rtgdk @goneall

**Comment by Yash-Varshney** [comment]
Also raising [this](https://github.com/Yash-Varshney/tools-python/issues/9) issue. A lot of points in this issue have been achieved but some are remaining, I will be raising issues for them ASAP.

**Comment by goneall** [comment]
@Yash-Varshney Thanks for the contributions and the PR!  

@zvr, @rtgdk & @pombredanne - if you could review and let me know if your OK with me merging.

File: spdx/relationship.py
Comment (RishabhBhatnagar) [inline spdx/relationship.py:102]: Rather than properties splitting the member variables every time and returning nth part,

It is preferable if you set these properties in the constructor itself, as the members (relationship and relationship_comment) are static for an object.
This will also block creation of an invalid Relationship which don't have enough parts (refA, relationship_type and, refB).

File: spdx/cli_tools/parser.py
Comment (RishabhBhatnagar) [inline spdx/cli_tools/parser.py:39]: [Optional] This if-elif-else ladder can be compressed too.

**Review by RishabhBhatnagar — CHANGES_REQUESTED** [review]
Overall, good work. Thank You for contributing to the SPDX community.
Although, I had a hard time scanning the significant changes because of petty format changes like changing single-quote to double-quote, formatting spaces, adding lines, etcetera.

If possible, segregate changes into two commits. One for significant changes and the other for formatting.

**Comment by rtgdk** [comment]
@Yash-Varshney @goneall I tried it out. And it works fine. I couldn't review the whole code since it takes too much time to review the real changes.

@Yash-Varshney Can you please respond to the comments made by @RishabhBhatnagar regarding the code structure? Overall, things look okay to merge if we have approvals from main reviewers.

File: examples/parse_json.py
Comment (Yash-Varshney) [reply examples/parse_json.py:7]: Thanks for your feedback Rishabh, I feel that `file` variable does more justice as it literally means the address of the path where the file is.

File: examples/parse_json.py
Comment (Yash-Varshney) [reply examples/parse_json.py:95]: @RishabhBhatnagar this parse_json (and all other files) are used for the cli_tool only. It actually doesn't make sense to use them without input files. 
Moreover, due to the introduction of cli_tool, these files won't be used .
Thanks

File: spdx/cli_tools/convertor.py
Comment (Yash-Varshney) [reply spdx/cli_tools/convertor.py:111]: This statement is basically to show the support for `rdf.xml` file to a reader. Ofc, it is unreachable but it actually tells the format supported :)

File: spdx/cli_tools/convertor.py
Comment (Yash-Varshney) [reply spdx/cli_tools/convertor.py:56]: During testing, the cases where they are none are working fine, please look out at the usage of the CLI-tools available in the readme. 
Thanks

File: spdx/cli_tools/convertor.py
Comment (Yash-Varshney) [reply spdx/cli_tools/convertor.py:125]: input variables can be anything (mostly strings ofc) but as a sanity check I deliberately converted it to string because if somehow there is a file say `12345.spdx` and the user wants to check it, it will break the code.

File: spdx/cli_tools/convertor.py
Comment (Yash-Varshney) [reply spdx/cli_tools/convertor.py:132]: good point, thanks. It could be improved further :)

**Comment by Yash-Varshney** [comment]
Thanks, @rtgdk  @zvr @goneall @RishabhBhatnagar for reviewing. Ik that due to formatting there are some redundant additions but it was needed to format the whole code once so that it won't be a problem in future. 
There are some small changes suggested by @RishabhBhatnagar, but I also believe that the code is working fine and it should be merged first and then few additions can be made.
Looking forward, I am thinking of adding a test for checking the use of `BLACK` formatter. 
Thank you :)

File: examples/parse_json.py
Comment (RishabhBhatnagar) [reply examples/parse_json.py:95]: 1. No user will make an error consciously. If he/she ends up not providing input, the program must terminate gracefully with an error message describing the usage.
Check out any of the examples in [tools-golang/examples](https://github.com/spdx/tools-golang/tree/master/examples) directory. Missing input describes the usage and terminates.
One more example: `mv file_name` when run in any unix-terminal will terminate stating "missing destination file".

2. Due to the introduction of cli_tool, which files won't be used?

Note: this is not a deal-breaker.

File: spdx/cli_tools/convertor.py
Comment (RishabhBhatnagar) [reply spdx/cli_tools/convertor.py:111]: subsuming case is when the outfile ends in ".xml".
If the outfile ends in ".xml", outfile_format is "xml"
If the outfile ends in ".rdf.xml", the expected outfile_format according to the code is "rdf" but the subsuming case will set the outfile_format to "xml".

This must be corrected.

File: spdx/cli_tools/convertor.py
Comment (RishabhBhatnagar) [reply spdx/cli_tools/convertor.py:125]: Even if the file name is "12345.spdx", the code won't break.

File: spdx/cli_tools/convertor.py
Comment (RishabhBhatnagar) [reply spdx/cli_tools/convertor.py:56]: I've tried running the convertor. But couldn't run it successfully for any of the ways defined in the [README.md](https://github.com/spdx/tools-python/blob/209a92563ea321403e0e8abc2308644367d73996/README.md) file.

The following are the ways I tried and the errors they've raised.
1. `convertor --infile file.rdf --outfile output.json` raised following error:
`Error: Invalid value for '--to' / '-t': invalid choice: undefined. (choose from json, rdf, yaml, xml, tag)`
2. `convertor --from rdf file.rdf --to json output.json` raised following error:
`spdx.parsers.builderexceptions.FileTypeError: INPUT FILETYPE NOT SUPPORTED. (only RDF and TAG format supported)`
3. `convertor -f rdf file.rdf -o output.json` raised following error:
`Error: Invalid value for '--to' / '-t': invalid choice: undefined. (choose from json, rdf, yaml, xml, tag)`

If you think I might've made some mistake while running the tool, you can check how I ran it in this [google colab file](https://colab.research.google.com/drive/1UUx2grrfII71fwCevYgbtCOted5qkS4t?usp=sharing)

Let me know any testcase for which the the aforementioned condition(line 56) doesn't hold True.

**Comment by goneall** [comment]
@Yash-Varshney The PR is showing that the branch is out of date with the base branch - can you update the branch and verify that it is still working.

@RishabhBhatnagar If your OK merging in its current state, can you create a separate issue for the outstanding issues and approve the PR?

**Comment by RishabhBhatnagar** [comment]
> @RishabhBhatnagar If your OK merging in its current state, can you create a separate issue for the outstanding issues and approve the PR?

@goneall, If you say so, I'll approve the PR. But I am not in the favor of merging it as I couldn't run the convertor for any of the described ways in the [readme file](https://github.com/spdx/tools-python/blob/efd0e59aeb5bd62379968e3e43e23a971bd414aa/README.md). 

Anyways, my review won't affect the merging of this PR as I am not a maintainer for this repository.

**Review by RishabhBhatnagar — APPROVED** [review]

**Pull request merged**
